:sparkles: Rewrite of copy & mirror tag solution

MathiasGruber opened the pull request:
# Pull Request
Re-write of the Copy & Mirror tag implementations. 

Previous PRs:
https://github.com/MathiasGruber/TheNinjaRPG/pull/421
https://github.com/MathiasGruber/TheNinjaRPG/pull/329

Issues: 
https://github.com/MathiasGruber/TheNinjaRPG/issues/328
https://github.com/MathiasGruber/TheNinjaRPG/issues/336

## License

By making this pull request, I confirm that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the Studie-Tech ApS organization has the copyright to use and modify my contribution for perpetuity.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced two new combat effects: **Copy** and **Mirror**.
    - **Copy** allows players to replicate beneficial effects from opponents.
    - **Mirror** enables players to reflect negative effects onto targets.
  - Updated effect ordering ensures these new mechanics integrate seamlessly during battle.

- **Improvements**
  - Enhanced effect processing logic to accommodate new effect types.
  - Simplified visibility logic for submit buttons across various forms, ensuring they are always displayed regardless of changes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MathiasGruber commented:
Did a few small adjustments to previous solutions

theeneon reviewed (dismissed):
While the text triggers incombat the abilities are not copied or mirrored

MathiasGruber commented:
> While the text triggers incombat the abilities are not copied or mirrored

I see why this is not working. The solution depends on how we want rounds to work. Say e.g. we set rounds=5 on copy - do we want it to copy positive effects the following rounds (i.e. also new positive effects after the tag was used), or do we want it to copy current positive effects and have them active for the next 5 rounds.

The latter is a lot simpler to implement I think.

theeneon commented:
> > While the text triggers incombat the abilities are not copied or mirrored
> 
> I see why this is not working. The solution depends on how we want rounds to work. Say e.g. we set rounds=5 on copy - do we want it to copy positive effects the following rounds (i.e. also new positive effects after the tag was used), or do we want it to copy current positive effects and have them active for the next 5 rounds.
> 
> The latter is a lot simpler to implement I think.

Its the latter that we would want to be implemented actually.

Phrosfire commented:
> > > While the text triggers incombat the abilities are not copied or mirrored
> > 
> > 
> > I see why this is not working. The solution depends on how we want rounds to work. Say e.g. we set rounds=5 on copy - do we want it to copy positive effects the following rounds (i.e. also new positive effects after the tag was used), or do we want it to copy current positive effects and have them active for the next 5 rounds.
> > The latter is a lot simpler to implement I think.
> 
> Its the latter that we would want to be implemented actually.

I switched copiedEffect.rounds = 1 to be copiedEffect.rounds = effect.rounds so that it would use the number of rounds from the tag itself, but it didn't change the outcome.  

I also added a return at the end of the else statement so we could know for sure that the else (where the effectcopy is happening) is triggering.  The return message is coming through, which means the foreach loop is being reached as well.  It just doesn't seem to be actually pushing the positive effects to the user.

MathiasGruber commented:
I'll work to fix it.. the reason is that we loop through the effects with forEach, but that does not account for new tags being added into the usersEffects

MathiasGruber commented:
@theeneon took a bit of a larger refactor, but I think it works as intended now

theeneon commented:
@MathiasGruber approve preemptively so that we can merge it when Vercel is done.

theeneon commented:
@MathiasGruber So tag works but is bugged, The effects are applying on every action which leads to a stacking issue Look below:

Mirror Test https://pastebin.com/iar3jJT1

As you can see consistent damage before the tag is applied after the tag is applied on every action the damage increases, in my side bar i had this:
![image](https://github.com/user-attachments/assets/d6a001a7-062b-47b9-996a-f71c4a144d27)

Instead of 30% debuff I had 90% debuff, when the tag was first used it was at 60% which means it's taking the 30% for 2 rounds and applying it all one round which leads to the 60%. I end my turn and it stacks 30% again which leads to the 90%. 

What should happen here is that the 30% is applied to me for the duration of the mirror tag but the effects shouldn't stack or be applied every action. 

Copy https://pastebin.com/DWeQYrKs

Now in this tag as you can see here, every time i moved the healing was applied which strengthens whats happening with mirror, that the effects that are copied are applying on each action. The status bar also shows this:

Movement 1: 
![image](https://github.com/user-attachments/assets/db5ee100-2e48-4d35-b200-17ee7d053376)

Movement 2:
![image](https://github.com/user-attachments/assets/e3e82b5d-47ce-49ee-a4f8-e9e66999d0e8)

Now this one is even more fun, Madcrep moved and I moved which also doubled the buff:
![image](https://github.com/user-attachments/assets/093bd94d-2305-4acc-b6bc-61f75241157a)

All of that ends with a beautiful one hit Madcrep takes 99772.93 damage 

So i believe once you're able to prevent the effects from applying on every action and only once per round like all other effects we're golden.

MathiasGruber commented:
> @MathiasGruber So tag works but is bugged, The effects are applying on every action which leads to a stacking issue Look below:
> 
> Mirror Test https://pastebin.com/iar3jJT1
> 
> As you can see consistent damage before the tag is applied after the tag is applied on every action the damage increases, in my side bar i had this: ![image](https://private-user-images.githubusercontent.com/48135705/423483151-d6a001a7-062b-47b9-996a-f71c4a144d27.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMjk5MTEsIm5iZiI6MTc0MjIyOTYxMSwicGF0aCI6Ii80ODEzNTcwNS80MjM0ODMxNTEtZDZhMDAxYTctMDYyYi00N2I5LTk5NmEtZjcxYzRhMTQ0ZDI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE3VDE2NDAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyYWRhYjM4ZDg3YmFiYmRmOTk1Yjk3ZTVhOGU5YjQ3Y2QxYjFkYTFlNDVkNWUxZGM0MTFkODViNGQzMDExNDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4wsgya94VxOXCG0xCzTuQ0NlcEIzsql4uL3d-nPy5Mg)
> 
> Instead of 30% debuff I had 90% debuff, when the tag was first used it was at 60% which means it's taking the 30% for 2 rounds and applying it all one round which leads to the 60%. I end my turn and it stacks 30% again which leads to the 90%.
> 
> What should happen here is that the 30% is applied to me for the duration of the mirror tag but the effects shouldn't stack or be applied every action.
> 
> Copy https://pastebin.com/DWeQYrKs
> 
> Now in this tag as you can see here, every time i moved the healing was applied which strengthens whats happening with mirror, that the effects that are copied are applying on each action. The status bar also shows this:
> 
> Movement 1: ![image](https://private-user-images.githubusercontent.com/48135705/423485629-db5ee100-2e48-4d35-b200-17ee7d053376.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMjk5MTEsIm5iZiI6MTc0MjIyOTYxMSwicGF0aCI6Ii80ODEzNTcwNS80MjM0ODU2MjktZGI1ZWUxMDAtMmU0OC00ZDM1LWIyMDAtMTdlZTdkMDUzMzc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE3VDE2NDAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwOTU0YTZiNzIxYWQ5ZmI2MzdjYmQ4NmI1Y2M5Y2MxMzM4ZWNjYTkyMGRlYjhhNjMzYzQ3NzlhNWMzMTMxZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.TvYzY6HGeFcc1wc8MrnUeYHfLLItEXB1D-NwhWnRS2M)
> 
> Movement 2: ![image](https://private-user-images.githubusercontent.com/48135705/423485683-e3e82b5d-47ce-49ee-a4f8-e9e66999d0e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMjk5MTEsIm5iZiI6MTc0MjIyOTYxMSwicGF0aCI6Ii80ODEzNTcwNS80MjM0ODU2ODMtZTNlODJiNWQtNDdjZS00OWVlLWE0ZjgtZTllNjY5OTlkMGU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE3VDE2NDAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBjMjhlZWRkMGMxZGNhYjI0MDA5MjU0Y2NjZmQ1MDFhM2EyN2QyMWViZDhlNmMzYzg1NDM4NGI2YTY1ZTIxZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jnKOT14Jht1jS6BNX0eXCg1wbW4Y1cjDXnCDRm1FwSM)
> 
> Now this one is even more fun, Madcrep moved and I moved which also doubled the buff: ![image](https://private-user-images.githubusercontent.com/48135705/423485845-093bd94d-2305-4acc-b6bc-61f75241157a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMjk5MTEsIm5iZiI6MTc0MjIyOTYxMSwicGF0aCI6Ii80ODEzNTcwNS80MjM0ODU4NDUtMDkzYmQ5NGQtMjMwNS00YWNjLWI2YmMtNjFmNzUyNDExNTdhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE3VDE2NDAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUyYzc5YWI1MmMxNWM2OTY4MjA0OTAwM2U5ZWNkMjI0YmI3YmQ4MDU3NTQ4MjcyZDRmOThlNDljOGQ0ZWUwZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.k85QbtLKwwbFiA7VwT2qI5eJRKCd7LHXI8lppeGucPk)
> 
> All of that ends with a beautiful one hit Madcrep takes 99772.93 damage
> 
> So i believe once you're able to prevent the effects from applying on every action and only once per round like all other effects we're golden.

Ah, I see, I think I've fixed it not so that they are not re-applied if already there. Works?
